packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -222,179 +240,273 @@ export default class RequestClient {
+          } catch (err) {
+            if (socketAbortController.signal.aborted) return
+            onFatalError(err)
+          }
         }
-        queuedRequest = queue.run(() => {
-          socket.open()
-          socket.send('resume', {})
 
-          return () => {}
-        })
-      })
-
-      eventManager.onRetry(file.id, () => {
-        // Only do the retry if the upload is actually in progress;
-        // else we could try to send these messages when the upload is still queued.
-        // We may need a better check for this since the socket may also be closed
-        // for other reasons, like network failures.
-        if (socket.isOpen) {
-          socket.send('pause', {})
-          socket.send('resume', {})
+        const pause = (newPausedState) => {
+          isPaused = newPausedState
+          if (socket) sendState()
+
+          if (newPausedState) {
+            // Remove this file from the queue so another file can start in its place.
+            socketAbortController?.abort?.() // close socket to free up the request for other uploads
+          } else {
+            // Resuming an upload should be queued, else you could pause and then
+            // resume a queued upload to make it skip the queue.
+            createWebsocket()
+          }
         }
-      })
 
-      eventManager.onRetryAll(file.id, () => {
-        // See the comment in the onRetry() call
-        if (socket.isOpen) {
-          socket.send('pause', {})
-          socket.send('resume', {})
+        const withErrorHandling = (fn, event) => {
+          try {
+            fn()
+          } catch (err) {
+            this.uppy.log(`Error while handling ${event} event for file ${file.id}: ${err.message}`, 'error')
+          }
         }
-      })
-
-      socket.on('progress', (progressData) => emitSocketProgress(this, progressData, file))
 
-      socket.on('error', (errData) => {
-        const { message } = errData.error
-        const error = Object.assign(new Error(message), {
-          cause: errData.error,
+        eventManager.onPause(file.id, (newPausedState) => withErrorHandling(() => pause(newPausedState), 'pause'))
+        eventManager.onPauseAll(file.id, () => withErrorHandling(() => pause(true), 'pause-all'))
+        eventManager.onResumeAll(file.id, () => withErrorHandling(() => pause(false), 'resume-all'))
+
+        eventManager.onFileRemove(file.id, () => withErrorHandling(() => {
+          socketSend('cancel')
+          socketAbortController?.abort?.()
+          this.uppy.log(`upload ${file.id} was removed`, 'info')
+          resolve()
+        }, 'file-removed'))
+
+        eventManager.onCancelAll(file.id, ({ reason } = {}) => withErrorHandling(() => {
+          if (reason === 'user') {
+            socketSend('cancel')
+          }
+          socketAbortController?.abort?.()
+          this.uppy.log(`upload ${file.id} was canceled`, 'info')
+          resolve()
+        }, 'cancel-all'))
```

Comment:
EventManager was already the wrong abstraction and now we are doubling down on it with another `withErrorHandling` abstraction. I already refactored this away in #4745.

Could use my refactored version and see if you even still need something like `withErrorHandling`? Also notice the if statements for certain events based on `capabilities`.

https://github.com/transloadit/uppy/blob/92995e847ebbed790c63032f56ebed660e2099c5/packages/%40uppy/companion-client/src/RequestClient.js#L286-L345

Reply:
Sure we could add the capabilities check, but im not sure it belongs in this PR. It’s not apparent to me how your code handles errors thrown in the event handlers, could you explain? For example if socket was null, it would crash the emit call in your code, no?

Reply:
It doesn't handle errors thrown yet. But since we're moving away from `EventManager`, it would help a lot if you do it now instead of making it even harder after merging this with more conflicts.

Reply:
If you dont think catching errors in an event handler is a good idea, we can discuss that, maybe it’s not. I just thought that it would be nice because np reported that if they cancel all uploads, and a single file throws an error in the handler, it prevents all other files from canceling too

Reply:
Sure i can refactor away eventmanager too. I just thought it was a new abstraction that was recently added so I didnt want to touch it

Reply:
I'm also thinking it's probably better to specifically check for something that might be undefined and model around that accordingly, rather than wrapping everything in a `try`/`catch`. imo `try`/`catch` should only be used for promises, not type errors.

Reply:
have now fixed this